runtime/metricsview/ast.go

```diff
@@ -783,21 +810,47 @@ func (a *AST) buildSpineSelect(alias string, spine *Spine, tr *TimeRange) (*Sele
 }
 
 // addTimeRange adds a time range to the given SelectNode's WHERE clause.
-func (a *AST) addTimeRange(n *SelectNode, tr *TimeRange) {
-	if tr == nil || tr.IsZero() || a.metricsView.TimeDimension == "" {
-		return
+func (a *AST) addTimeRange(n *SelectNode, tr *TimeRange) error {
+	if tr == nil || tr.IsZero() || (a.metricsView.TimeDimension == "" && tr.TimeDimension == "") {
+		return nil
 	}
 
 	// Since resolving time ranges may require contextual info (like watermarks), the upstream caller is responsible for resolving them.
 	if tr.Start.IsZero() && tr.End.IsZero() {
 		panic("ast received a non-empty, unresolved time range")
 	}
 
-	expr, args := a.sqlForTimeRange(a.metricsView.TimeDimension, tr.Start, tr.End)
+	timeDimExpr, err := a.getTimeDimensionExpression(tr)
+	if err != nil {
+		return err
+	}
+
+	expr, args := a.sqlForTimeRange(timeDimExpr, tr.Start, tr.End)
 	n.TimeWhere = &ExprNode{
 		Expr: expr,
 		Args: args,
 	}
+
+	return nil
+}
+
+func (a *AST) getTimeDimensionExpression(tr *TimeRange) (string, error) {
```

Comment:
nits: maybe add a little docstring? it's a big file and most functions in it have a little docstring